File: frontend/src/api/nativeTools.ts
import type {
  NativeToolArtifact,
  NativeToolDefinition,
  NativeToolRunResult,
  SystemSettings,
} from "../lib/nativeToolTypes";
import { NATIVE_TOOLS_STATIC } from "../lib/nativeToolTypes";

const ALL_TOOL_IDS = NATIVE_TOOLS_STATIC.map((t) => t.toolId);

const API_BASE = import.meta.env.VITE_API_BASE ?? "";

function apiUrl(path: string): string {
  return `${API_BASE}${path}`;
}

async function parseError(res: Response): Promise<string> {
  try {
    const body = (await res.json()) as Record<string, unknown>;
    if (typeof body.detail === "string") return body.detail;
    if (typeof body.error === "string") return body.error;
    return `请求失败（HTTP ${res.status}）`;
  } catch {
    return res.statusText || `请求失败（HTTP ${res.status}）`;
  }
}

export async function fetchNativeTools(): Promise<NativeToolDefinition[]> {
  try {
    const res = await fetch(apiUrl("/api/tools"));
    if (!res.ok) throw new Error(await parseError(res));
    return res.json();
  } catch {
    return NATIVE_TOOLS_STATIC;
  }
}

export async function fetchNativeToolArtifacts(): Promise<NativeToolArtifact[]> {
  try {
    const res = await fetch(apiUrl("/api/tools/artifacts"));
    if (!res.ok) return [];
    const data = (await res.json()) as { artifacts: NativeToolArtifact[] };
    return data.artifacts ?? [];
  } catch {
    return [];
  }
}

export async function fetchNativeArtifactContent(
  jobId: string,
  filename: string,
): Promise<{ content: string; contentType: string }> {
  const res = await fetch(apiUrl(`/api/tools/artifacts/${encodeURIComponent(jobId)}/${filename}`));
  if (!res.ok) throw new Error(await parseError(res));
  const data = await res.json();
  return { content: data.content, contentType: data.contentType };
}

export async function convertMarkdown(form: FormData): Promise<NativeToolRunResult> {
  const res = await fetch(apiUrl("/api/tools/markdown/convert"), { method: "POST", body: form });
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

export async function scanFolder(body: {
  target_path: string;
  scan_mode: string;
  dry_run: boolean;
  max_depth: number;
}): Promise<NativeToolRunResult> {
  const res = await fetch(apiUrl("/api/tools/folder-clean/scan"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

export async function generateGitHubResume(body: {
  repo_url: string;
  language: string;
  role_target: string;
}): Promise<NativeToolRunResult> {
  const res = await fetch(apiUrl("/api/tools/github-resume/generate"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

export async function queryRagStub(body: {
  document_path: string;
  question: string;
  top_k: number;
  model: string;
}): Promise<NativeToolRunResult> {
  const res = await fetch(apiUrl("/api/tools/rag/query"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

export async function downloadBilibiliStub(body: {
  url: string;
  output_dir: string;
  quality: string;
}): Promise<NativeToolRunResult> {
  const res = await fetch(apiUrl("/api/tools/bilibili/download"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

export async function runQuickForgeStub(body: {
  script: string;
  args: string[];
}): Promise<NativeToolRunResult> {
  const res = await fetch(apiUrl("/api/tools/quickforge/run"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

export async function fetchSystemSettings(): Promise<SystemSettings> {
  try {
    const res = await fetch(apiUrl("/api/settings"));
    if (!res.ok) throw new Error(await parseError(res));
    return res.json();
  } catch {
    return {
      model: { provider: "demo", modelName: "demo", apiKeyConfigured: false, demoMode: true },
      storage: { workspaceRoot: "workspace", artifactRoot: "tool_artifacts", maxArtifactSizeMb: 50 },
      tool: {
        enabledTools: ALL_TOOL_IDS,
        defaultOutputDirectory: "tool_output",
        dryRunDefault: true,
      },
      runtime: {
        maxIterations: 8,
        plannerTimeoutSeconds: 20,
        toolTimeoutSeconds: 30,
        statusPollingIntervalMs: 2000,
      },
    };
  }
}

export async function saveSystemSettings(patch: Partial<SystemSettings>): Promise<SystemSettings> {
  const res = await fetch(apiUrl("/api/settings"), {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(patch),
  });
  if (!res.ok) throw new Error(await parseError(res));
  return res.json();
}

File: backend/src/api/routes/native_tools.py
"""Native Tool Hub API — built-in modules with artifact output."""

from __future__ import annotations

from pathlib import Path
from typing import Any

from fastapi import APIRouter, File, Form, HTTPException, UploadFile
from pydantic import BaseModel, Field

from src.native_tools.artifact_store import NativeArtifactStore
from src.native_tools.folder_clean import scan_folder
from src.native_tools.github_resume import generate_resume
from src.native_tools.markdown import convert_markdown
from src.native_tools.registry import get_native_tool, list_native_tools
from src.native_tools.stubs import stub_bilibili_download, stub_quickforge_run, stub_rag_query

router = APIRouter(prefix="/api/tools", tags=["native-tools"])


def _store() -> NativeArtifactStore:
    return NativeArtifactStore()


class FolderCleanRequest(BaseModel):
    target_path: str
    scan_mode: str = "all"
    dry_run: bool = True
    max_depth: int = Field(default=8, ge=1, le=32)


class GitHubResumeRequest(BaseModel):
    repo_url: str
    language: str = "zh"
    role_target: str = ""


class RagQueryRequest(BaseModel):
    document_path: str
    question: str
    top_k: int = Field(default=5, ge=1, le=20)
    model: str = "ollama"


class BilibiliDownloadRequest(BaseModel):
    url: str
    output_dir: str = "./downloads"
    quality: str = "1080p"


class QuickForgeRunRequest(BaseModel):
    script: str
    args: list[str] = Field(default_factory=list)


@router.get("/native")
async def list_tools_native():
    store = NativeArtifactStore()
    tools = list_native_tools()
    recent = store.list_recent(limit=30)
    for tool in tools:
        tool["recentArtifacts"] = [a for a in recent if a.get("toolId") == tool["toolId"]][:5]
    return tools


@router.get("/native/{tool_id}")
async def get_tool_native(tool_id: str):
    tool = get_native_tool(tool_id)
    if tool is None:
        raise HTTPException(status_code=404, detail=f"Unknown tool: {tool_id}")
    recent = NativeArtifactStore().list_recent(limit=20)
    tool["recentArtifacts"] = [a for a in recent if a.get("toolId") == tool_id]
    return tool


@router.get("/artifacts")
async def list_tool_artifacts():
    return {"artifacts": NativeArtifactStore().list_recent(limit=100)}


@router.get("/artifacts/{job_id}/{filename:path}")
async def get_tool_artifact(job_id: str, filename: str) -> dict[str, Any]:
    store = NativeArtifactStore()
    try:
        content = store.read_text(job_id, filename)
    except FileNotFoundError:
        raise HTTPException(status_code=404, detail=f"Artifact not found: {filename}")
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))
    lower = filename.lower()
    if lower.endswith((".md", ".markdown")):
        content_type = "text/markdown"
    elif lower.endswith(".json"):
        content_type = "application/json"
    else:
        content_type = "text/plain"
    return {"name": filename, "content": content, "contentType": content_type, "jobId": job_id}


@router.post("/markdown/convert")
async def markdown_convert(
    file: UploadFile = File(...),
    source_type: str = Form(""),
    output_filename: str = Form(""),
):
    raw = await file.read()
    filename = file.filename or "upload.txt"
    return await convert_markdown(
        _store(),
        file_bytes=raw,
        filename=filename,
        source_type=source_type or Path(filename).suffix.lstrip("."),
        output_filename=output_filename,
    )


@router.post("/folder-clean/scan")
async def folder_clean_scan(body: FolderCleanRequest):
    return await scan_folder(
        _store(),
        target_path=body.target_path,
        scan_mode=body.scan_mode,
        dry_run=body.dry_run,
        max_depth=body.max_depth,
    )


@router.post("/github-resume/generate")
async def github_resume_generate(body: GitHubResumeRequest):
    return await generate_resume(
        _store(),
        repo_url=body.repo_url,
        language=body.language,
        role_target=body.role_target,
    )


@router.post("/rag/query")
async def rag_query(body: RagQueryRequest):
    return await stub_rag_query(
        _store(),
        document_path=body.document_path,
        question=body.question,
        top_k=body.top_k,
        model=body.model,
    )


@router.post("/bilibili/download")
async def bilibili_download(body: BilibiliDownloadRequest):
    return await stub_bilibili_download(
        _store(),
        url=body.url,
        output_dir=body.output_dir,
        quality=body.quality,
    )


@router.post("/quickforge/run")
async def quickforge_run(body: QuickForgeRunRequest):
    return await stub_quickforge_run(_store(), script=body.script, args=body.args)
